Support environment variables in docker-container driver

sirlatrom opened the pull request:
Fixes #169

Example usage:

```bash
docker buildx create --driver-opt env.http_proxy=$http_proxy --driver-opt env.https_proxy=$https_proxy --driver-opt '"env.no_proxy='$no_proxy'"'
```

---

Original suggestion:

```bash
docker buildx create --driver-env http_proxy=$http_proxy --driver-env https_proxy=$https_proxy --driver-env '"no_proxy='$no_proxy'"'
```

tonistiigi commented:
Maybe we could reuse `DriverOpt` for this as they would only be present in certain drivers.

sirlatrom commented:
> Maybe we could reuse `DriverOpt` for this as they would only be present in certain drivers.

The function for passing in values for DriverOpt does not allow multiple values for the same key; hence the new field.

sirlatrom commented on commands/create.go:
```diff
@@ -215,7 +221,7 @@ func csvToMap(in []string) (map[string]string, error) {
 		for _, v := range fields {
 			p := strings.SplitN(v, "=", 2)
-			if len(p) != 2 {
+			if len(p) < 2 {
```
I should revert this line if the separate `DriverEnv` field is accepted.

tonistiigi commented on commands/create.go:
```diff
@@ -215,7 +221,7 @@ func csvToMap(in []string) (map[string]string, error) {
+			if len(p) < 2 {
 				return nil, errors.Errorf("invalid value %q, expecting k=v", v)
 			}
 			m[p[0]] = p[1]
```
We could do it either by changing the internal data type to `map[string][]string` or requesting the input as `env.FOO=bar`

sirlatrom replied:
Done in ac566dda9c22f0c1bbffa7d6c85e01eb58ee4f3e

tonistiigi commented on commands/create.go:
```diff
@@ -215,10 +215,10 @@ func csvToMap(in []string) (map[string]string, error) {
 				return nil, errors.Errorf("invalid value %q, expecting k=v", v)
 			}
-			m[p[0]] = p[1]
+			m[p[0]] = strings.Join(p[1:], "=")
```
What is this change for?

sirlatrom replied:
Environment variable values may contain `=` (and, frankly, `,`, which leads me to think the current approach will not do).

tonistiigi replied:
It uses `SplitN` so it will never return more than 2 results.

sirlatrom replied:
Fair point, I'll revert this change :+1:

tonistiigi commented on driver/docker-container/factory.go:
```diff
@@ -39,14 +41,20 @@ func (f *factory) New(ctx context.Context, cfg driver.InitConfig) (driver.Driver
+		case k == "image":
 			d.image = v
+		case strings.HasPrefix(k, "env."):
+			envName := k[4:]
```
`strings.TrimPrefix` is cleaner

sirlatrom replied:
Fixed in 046a3b0939a7c7b2c036697a44380b4d7f55151f.

tonistiigi commented on driver/docker-container/factory.go:
```diff
@@ -39,14 +41,20 @@ func (f *factory) New(ctx context.Context, cfg driver.InitConfig) (driver.Driver
 			d.image = v
+		case strings.HasPrefix(k, "env."):
+			envName := k[4:]
+			if len(envName) == 0 {
```
nit: `envName == ""`

sirlatrom replied:
Fixed in 046a3b0939a7c7b2c036697a44380b4d7f55151f.

tonistiigi commented:
@sirlatrom Please squash as well. And avoid putting username mentions in the commit message. I get pinged every time it is pushed or rebased in a fork.

sirlatrom commented:
> @sirlatrom Please squash as well. And avoid putting username mentions in the commit message. I get pinged every time it is pushed or rebased in a fork.

Squashed and rebased, thanks for reviewing!

spejic88 commented:
when I use `docker buildx create --driver-opt env.http_proxy=$http_proxy --driver-opt env.https_proxy=$https_proxy --driver-opt '"env.no_proxy='$no_proxy'"'`
to set proxy I get `invalid driver option env.http_proxy for docker-container driver`

tavlima commented:
Hey, @sirlatrom, nice work here. Do you think this could be "ported" to the `remote` driver as well?

sirlatrom commented:
> Hey, @sirlatrom, nice work here. Do you think this could be "ported" to the `remote` driver as well?

Feel free to borrow what you need from this PR. I'm not currently in a position where I can make additional contributions to this project.

tavlima commented:
Yeah, I'm aware of what the `remote` driver is, @tonistiigi, but it is just not clear to me yet how the environments (client env, build args, remote env, etc) "interact".

My buildkit nodes run in an environment where all the egress traffic must go through a proxy and I had a hard time getting it to work.

Perhaps something was wrong with my experiments, but AFAIR just making sure the `buildkit` remote nodes had, themselves, the environment variables set (`{NO,HTTP,HTTPS}_PROXY` and their lowercase counterparts) didn't fix the issues while pulling public images. IIRC, I had to also invoke `buildx` like `docker buildx build --builder my_remote_builder --build-arg HTTPS_PROXY=... --build-arg HTTP_PROXY=...`.

Is that expected?

tonistiigi commented:
Yes, that is expected. To configure proxy for the internal containers created by `RUN` steps, they need to be passed with build-arg. They do not get inherited from the daemon environment.

tavlima commented:
So, yeah, that is what I was hoping could be addressed in a similar way as this PR. I'd rather not require all my clients to know about the specifics of the cluster environment.

In my case, my team is expected to maintain a cluster with tens of buildkit instances, and I'd rather not have our whole org worrying about what is the right proxy configuration to fetch image XYZ, especially because our environment requires dozens of CIDRs and wildcard domains to be included in the `NO_PROXY` envvar.

Completely abstracting that away from the clients would be ideal. Doing this setup just once (like when "registering" the remote builder in each client) would be an improvement. But having to include `--build-arg` for all proxy environment variables, every time one calls our `--builder`? That will be a hard sell...

tavlima commented:
FWIW, I just realized what confused me in my experiments: the daemon environment variables are inherited, at least for directives like `FROM`. But, just like you said, @tonistiigi... 

> To configure proxy for the internal containers created by RUN steps, they need to be passed with build-arg

It is really odd that we need to have the same configuration in both places (daemon env and `--build-arg`) in order to build single image (assuming proxies are required for whatever image/command is used in both `FROM`/`RUN`)
